### PR TITLE
Chromap Module

### LINKED ISSUE #540: new module: CHROMAP
I think it would be good to have a module for [Chromap](https://bioconda.github.io/recipes/chromap/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by mahesh-panchal**
Addition of the chromap module.

## PR checklist

Closes #540

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by mahesh-panchal** [comment]
Does anyone with more experience of ATAC-seq or CHIP-seq know if any of the optional input files are sample specific? 
i.e., should any of the input files be added to the meta map?

**Comment by mahesh-panchal** [comment]
Also, should I compress all the outputs?

File: modules/chromap/index/main.nf [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    } else {
+        container "quay.io/biocontainers/mulled-v2-1f09f39f20b1c4ee36581dc81cc323c70e661633:2cad7c5aa775241887eff8714259714a39baf016-0"
+    }
+
+    input:
+    path fasta
+
+    output:
+    path "*.index"       , emit: index
+    path "*.version.txt" , emit: version
```
Comment (drpatelh) [inline modules/chromap/index/main.nf:28]:
```suggestion
    path "*.index"      , emit: index
    path "*.version.txt", emit: version
```

File: modules/chromap/index/main.nf [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix = fasta.baseName
```
Comment (drpatelh) [inline modules/chromap/index/main.nf:32]:
```suggestion
    def software = getSoftwareName(task.process)
    def prefix   = fasta.baseName
```

File: modules/chromap/index/main.nf [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    """
+    chromap -i $options.args \\
+        -t ${task.cpus} \\
```
Comment (drpatelh) [inline modules/chromap/index/main.nf:35]:
```suggestion
        -t $task.cpus \\
```

File: modules/chromap/chromap/main.nf [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            -x $index \\
+            -r $fasta \\
+            -1 ${reads.join(',')} \\
+            -o ${prefix}.${file_extension}
+        echo "$VERSION" > ${software}.version.txt
+        """ + compression_cmds
+    } else {
+        """
+        chromap ${args.join(' ')} \\
+            -t ${task.cpus} \\
```
Comment (drpatelh) [inline modules/chromap/chromap/main.nf:84]:
```suggestion
            -t $task.cpus \\
```

File: tests/config/pytest_software.yml [hunk 5]
```diff
@@ -0,0 +1,748 @@
+abacas:
+  - software/abacas/**
```
Comment (drpatelh) [inline tests/config/pytest_software.yml:2]: This file needs to be removed from the PR.

File: tests/modules/chromap/chromap/main.nf [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    // Test single-end and gz compressed output
+
+    fasta = file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
+    input = [
+        [ id:'test', single_end:true ], // meta map
+        [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true) ]
+    ]
+
+    CHROMAP_INDEX( fasta )
```
Comment (drpatelh) [inline tests/modules/chromap/chromap/main.nf:19]:
```suggestion
    CHROMAP_INDEX ( fasta )
```

File: tests/modules/chromap/chromap/main.nf [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    fasta = file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
+    input = [
+        [ id:'test', single_end:false ], // meta map
+        [
+            file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
+            file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true)
+        ]
+    ]
+
+    CHROMAP_INDEX( fasta )
```
Comment (drpatelh) [inline tests/modules/chromap/chromap/main.nf:44]:
```suggestion
    CHROMAP_INDEX ( fasta )
```

File: tests/modules/chromap/chromap/main.nf [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
```
Comment (drpatelh) [inline tests/modules/chromap/chromap/main.nf:69]:
```suggestion
    CHROMAP_INDEX ( fasta )
```

**Review by drpatelh — COMMENTED** [review]
😍 Minor changes @mahesh-panchal ! Thank you :)

File: tests/config/pytest_software.yml (hunk 5 above)
Comment (mahesh-panchal) [reply tests/config/pytest_software.yml]: Removed file.

File: modules/chromap/chromap/main.nf [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            -o ${prefix}.bam ${prefix}.${file_extension}
+        rm ${prefix}.${file_extension}
+
+        samtools --version 2>&1 | sed 's/^.*samtools //; s/Using.*\$//' > ${software}.version.txt
+        """
+    }
+    if (meta.single_end) {
+        """
+        chromap ${args.join(' ')} \\
+            -t ${task.cpus} \\
```
Comment (drpatelh) [inline modules/chromap/chromap/main.nf:74]:
```suggestion
            -t $task.cpus \\
```

**Review by drpatelh — APPROVED** [review]
Love the optional logic you have incorporated here to deal with different input/output options.

**Pull request merged**